### DISCUSSION TITLE
Is it possible to use FPP to generate standalone packet definitions?

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#1963] [Q&A] [answered]
In my project, we have another embedded system far too small to use F', and we are looking at unifying the TM/TC regimes for both.

The my F'-centric colleagues are very happy with FPP, and although we will move away from the inbuilt F' TM/TC regime, we would like to continue to use FPP with our F' deployment, and extend it to specifying TM/TC for the other system, to build packet definitions.

So I whipped up this quick demo to try to extract an XML dictionary:
```
module Demo {
    
    enum TransitionCommand : U8 {
        STOP      = 0
        START     = 1
        RESET     = 2
        CALIBRATE = 3
    }

    enum SubSystem : U8 {
        SS1 = 0
        SS2 = 1
    }

    array CanPayload = [8] U8

    passive component ModuleDemo {

        guarded command SEND_MESSAGE( \
            sys_id:        U8
            subsystem:     SubSystem
            f_code:        U8
            c_id:          U32
            value_buffer:  CanPayload
            data_length:   U8
        ) opcode 0

        guarded command UPDATE_STATE_REQUEST( \
            sys_id:        U8
        ) opcode 1

        guarded command SET_STATE( \
            sys_id: U8
            target_state: TransitionCommand
        ) opcode 2
    }

    instance demoMod: ModuleDemo base id 0x0100
}
```

Trying to convert this to XML, however, gives this error:
```
fpp-to-xml
/path/to/MyTest.fpp: 21.5
    passive component ModuleDemo {
    ^
error: component with command specifiers must have command recv port
```

I added the `recv` port, but this results in:
```
fpp-to-xml
/path/to/MyTest.fpp: 23.9
        command recv port cmdIn
        ^
error: undefined symbol Fw
```

Is there a way around the requirement for an actual F' installation, or am I onto a loser here?

**Comment by bocchino** [comment, marked as answer]
fpp-check just requires that various symbols be defined. It does not enforce any particular definition. For example, if you add
```
module Fw {
  port Cmd
  port CmdReg
  port CmdResponse
}
```
then your example should pass through fpp-check. We use this approach frequently for unit testing FPP, or for constructing small instructional examples.

To generate meaningful C++ code with fpp-to-cpp, you must use the port models provided in Fw/Cmd in the F Prime repo.

**Reply by JackNWhite** [reply]
Thanks, that's really helpful.
